ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
 import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
- * The underlying file is backed by a log abstraction. This allows writes as append only operations.
- * For pre-allocated files, this tracks the end of valid file to ensure appends happen correctly.
- * It provides ability to read from arbitrary offset into the file. It can also provide a static view
- * of the log for a given set of offset,size pairs.
+ * The Log is an abstraction over a group of files that are loaded as {@link LogSegment} instances. It provides a
+ * unified view of these segments and provides a way to interact and query different properties of this group of
+ * segments.
  */
-public class Log implements Read, Write {
-
-  private AtomicLong currentWriteOffset;
-  private final FileChannel fileChannel;
-  private final File file;
+public class Log implements Write {
+  private final String dataDir;
   private final long capacityInBytes;
-  private static final String Log_File_Name = "log_current";
-  private Logger logger = LoggerFactory.getLogger(getClass());
   private final StoreMetrics metrics;
+  private final AtomicLong remainingUnallocatedCapacity;
+  private final ConcurrentSkipListMap<String, LogSegment> segmentsByName =
+      new ConcurrentSkipListMap<>(LogSegmentNameHelper.COMPARATOR);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private LogSegment activeSegment;
 
-  public Log(String dataDir, long capacityInBytes, StoreMetrics metrics)
+  /**
+   * Create a Log instance
+   * @param dataDir the directory where the segments of the log need to be created.
+   * @param totalCapacityInBytes the total capacity of this log.
+   * @param segmentCapacityInBytes the capacity of a single segment in the log.
+   * @param metrics the {@link StoreMetrics} instance to use.
+   * @throws IOException if there is any I/O error loading the segment files.
+   * @throws IllegalArgumentException if {@code totalCapacityInBytes} or {@code segmentCapacityInBytes} <= 0 or if
+   * {@code totalCapacityInBytes} > {@code segmentCapacityInBytes} and {@code totalCapacityInBytes} is not a perfect
+   * multiple of {@code segmentCapacityInBytes}.
+   */
+  public Log(String dataDir, long totalCapacityInBytes, long segmentCapacityInBytes, StoreMetrics metrics)
       throws IOException {
-    file = new File(dataDir, Log_File_Name);
-    if (!file.exists()) {
-      // if the file does not exist, preallocate it
-      Utils.preAllocateFileIfNeeded(file, capacityInBytes);
+    if (totalCapacityInBytes <= 0 || segmentCapacityInBytes <= 0) {
+      throw new IllegalArgumentException(
+          "One of totalCapacityInBytes [" + totalCapacityInBytes + "] or " + "segmentCapacityInBytes ["
+              + segmentCapacityInBytes + "] is <=0");
+    }
+    long segmentCapacity = Math.min(totalCapacityInBytes, segmentCapacityInBytes);
```

Comment:
Not always. In case someone is not interested in segmentation, they would set the config to `Long.MAX_VALUE`.